Dockerfile

```
@@ -0,0 +1,77 @@
+FROM node:20-slim AS base
+ENV PNPM_HOME="/pnpm"
+ENV PATH="$PNPM_HOME:$PATH"
+RUN corepack enable pnpm
+
+FROM base AS base-deps
+WORKDIR /home/node/repo
+
+COPY ./package.json ./pnpm-*.yaml ./
+COPY ./lib/package.json ./lib/
+RUN --mount=type=cache,id=pnpm,target=/pnpm/store pnpm install --frozen-lockfile
+
+
+
+# Express
+FROM base AS express-deps
+WORKDIR /home/node/repo
+COPY --from=base-deps /home/node/repo ./
+
+COPY ./examples/express/package.json ./examples/express/
+RUN --mount=type=cache,id=pnpm,target=/pnpm/store pnpm install --frozen-lockfile
+
+
+FROM base AS express-build
+WORKDIR /home/node/repo
+COPY --from=express-deps /home/node/repo ./
+COPY . /home/node/repo
+
+RUN pnpm --filter !esi-next run build
+RUN pnpm deploy --filter esi-express --prod /home/node/esi-express
+
+FROM base AS express-prod
+COPY --from=express-build /home/node/esi-express /home/node/esi-express
+WORKDIR /home/node/esi-express
+USER node
+EXPOSE 3000
+CMD [ "node", "dist/server.js" ]
+
+
+
+# Nextjs
+FROM base AS next-deps
+WORKDIR /home/node/repo
+COPY --from=base-deps /home/node/repo ./
+
+COPY ./examples/next/package.json ./examples/next/
+RUN --mount=type=cache,id=pnpm,target=/pnpm/store pnpm install --frozen-lockfile
+
+
+FROM base AS next-build
+WORKDIR /home/node/repo
+COPY --from=next-deps /home/node/repo ./
+COPY . /home/node/repo
+
+ENV NEXT_TELEMETRY_DISABLED 1
+RUN pnpm --filter !esi-express run build
+RUN pnpm deploy --filter esi-next --prod /home/node/esi-next
+
+
+FROM base AS next-prod
+WORKDIR /home/node/esi-next
+
+ENV NODE_ENV production
+ENV NEXT_TELEMETRY_DISABLED 1
+
+# Set the correct permission for prerender cache
+RUN mkdir .next
+RUN chown node:node .next
+COPY --from=next-build --chown=node:node /home/node/esi-next/public ./public
+COPY --from=next-build --chown=node:node /home/node/esi-next/node_modules ./node_modules
+COPY --from=next-build --chown=node:node /home/node/esi-next/.next/standalone ./
+COPY --from=next-build --chown=node:node /home/node/esi-next/dist ./dist
+COPY --from=next-build --chown=node:node /home/node/esi-next/.next/static ./.next/static
```

```suggestion
COPY --from=next-build --chown=1001:1001 /home/node/esi-next/public ./public
COPY --from=next-build --chown=1001:1001 /home/node/esi-next/node_modules ./node_modules
COPY --from=next-build --chown=1001:1001 /home/node/esi-next/.next/standalone ./
COPY --from=next-build --chown=1001:1001 /home/node/esi-next/dist ./dist
COPY --from=next-build --chown=1001:1001 /home/node/esi-next/.next/static ./.next/static
```

See https://github.com/api-platform/api-platform/issues/2412.

---

It seems to break only when using --chown and --link in the same COPY command so not strictly necessary here.
https://github.com/moby/buildkit/issues/2987
If you prefer to use uid/gui, we should use 1000:1000 as it is how it defined in the [node image](https://github.com/nodejs/docker-node/blob/e5ffb12d5d0e763d68e2766e8345faeb53d29a0f/20/bookworm-slim/Dockerfile#L3)

---

Actually, it's recommended to use a dedicated `nextjs` user as done in API Platform. But let's keep it as-is for now.

---

I've seen that, I don't know why Next create and use its own user in another node group, maybe to make their user's rights stricter?

Is was added [at this time](https://github.com/vercel/next.js/commit/d49f978a17673804311d8299a3af01792b2775b5#diff-1b39089dfd2ececdb6d87ca6169697e443df4642acae9dbeb7fc2726d6fc5c3fR29) but there is no explanation.